third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -103,34 +105,183 @@ Value TargetInfo::loadDShared(RewriterBase &rewriter, Location loc, Value ptr,
 
 Value TargetInfo::shuffleXor(RewriterBase &rewriter, Location loc, Value val,
                              int i) const {
-  return LLVM::AMD::shuffleXor(loc, rewriter, val, i);
+  return LLVM::AMD::shuffleXor(loc, rewriter, val, i, getISAFamily());
 }
 
 Value TargetInfo::shuffleUp(RewriterBase &rewriter, Location loc, Value val,
                             int i) const {
-  return LLVM::AMD::shuffleUp(loc, rewriter, val, i);
+  return LLVM::AMD::shuffleUp(loc, rewriter, val, i, getISAFamily());
 }
 
 Value TargetInfo::shuffleIdx(RewriterBase &rewriter, Location loc, Value val,
                              int i) const {
-  return LLVM::AMD::shuffleIdx(loc, rewriter, val, i);
+  return LLVM::AMD::shuffleIdx(loc, rewriter, val, i, getISAFamily());
 }
 
 Value TargetInfo::shuffleIdx(RewriterBase &rewriter, Location loc, Value val,
                              Value i) const {
-  return LLVM::AMD::shuffleIdx(loc, rewriter, val, i);
+  return LLVM::AMD::shuffleIdx(loc, rewriter, val, i, getISAFamily());
 }
 
 Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
                             ModuleOp moduleOp, int axis) const {
   return LLVM::AMD::llGetPid(loc, rewriter, moduleOp, axis);
 }
 
+static inline Type castToInt(RewriterBase &rewriter, Location loc, Value &val,
+                             Type valType, unsigned bits) {
+  unsigned originalBits = valType.getIntOrFloatBitWidth();
+  Type actualType = valType;
+
+  if (!valType.isIntOrIndex()) {
+    val = bitcast(val, int_ty(originalBits));
+    actualType = int_ty(originalBits);
+  }
+
+  if (originalBits < bits) {
+    val = sext(int_ty(bits), val);
+    actualType = int_ty(bits);
+  }
+
+  return actualType;
+}
+
+static inline void castFromInt(RewriterBase &rewriter, Location loc, Value &val,
```

Comment:
Prefer to return the casted `val` instead of mutating it inplace, esp. here we don't have other things to return.